functional_programming_tests/test/with_dependency_injection.cpp

```diff
@@ -0,0 +1,528 @@
+            response->path = std_msgs::msg::UInt8MultiArray();
+            return;
+          }
+          auto const start =
+              Position{request->start.data[0], request->start.data[1]};
+          auto const goal =
+              Position{request->goal.data[0], request->goal.data[1]};
+
+          // Generate the path
+          auto const path = generate_global_path(start, goal);
+
+          // Start populating the response message
+          auto response_path = std_msgs::msg::UInt8MultiArray();
+
+          if (path.has_value()) {
+            response_path.layout.dim.resize(
+                3, std_msgs::msg::MultiArrayDimension());
+
+            response_path.layout.dim[0].label = "rows";
+            response_path.layout.dim[0].size = path.value().size();
+            response_path.layout.dim[0].stride = path.value().size() * 2;
+
+            response_path.layout.dim[1].label = "columns";
+            response_path.layout.dim[1].size = 2;
+            response_path.layout.dim[1].stride = 1;
+
+            response_path.layout.dim[2].label = "channel";
+            response_path.layout.dim[2].size = 1;
+            response_path.layout.dim[2].stride = 1;
+
+            // Start pushing back the path only if there is one
+            if (path.value().size() > 0) {
+              for (auto const& position : path.value()) {
+                response_path.data.push_back(position.x);
+                response_path.data.push_back(position.y);
+              }
+            }
+          }
+
+          response->success.data = path.has_value();
+          response->path = response_path;
+        });
+  }
+
+ private:
+  /**
+   * @brief Extracts costmap from message and sets it
+   * @param costmap The costmap message
+   * @return True if the costmap was set successfully, false otherwise
+   */
+  bool costmap_setter(
+      std_msgs::msg::UInt8MultiArray const& costmap) {  // Action
+    // Get the costmap from a ros topic
+    // Check that map layout makes sense
+    if ((costmap.layout.dim[0].size * costmap.layout.dim[1].size) !=
+        costmap.layout.dim[0].stride) {
+      mw_->log_error("COSTMAP DIMENSIONS AND STRIDE INCONSISTENT!!");
+      return false;
+    }
+    if (costmap.layout.dim[0].stride != costmap.data.size()) {
+      mw_->log_error("COSTMAP LENGTH AND STRIDE INCONSISTENT!!");
+      return false;
+    }
+    auto const begin = std::begin(costmap.data);
+    // Populate the map
+    for (size_t row = 0; row < costmap.layout.dim[0].size; row++) {
+      map_.get_data().push_back(
+          {begin + row * costmap.layout.dim[1].size,
+           begin + (row + 1) * costmap.layout.dim[1].size});
+    }
+    return true;
+  }
+
+  /**
+   * @brief Generates a path from start to goal
+   * @param start The start position
+   * @param goal The goal position
+   * @return The path from start to goal if it exists, std::nullopt otherwise
+   */
+  std::optional<Path> generate_global_path(
```

Comment:
Should the return type be `Path` or `std::optional<Path>`? Does it defeat the purpose to have it be optional?

Reply:
If the code can fail to produce an output, returning a `std::optional` is a great way to communicate that, at least until we have `std::expected`